src/main/java/org/openstreetmap/atlas/checks/base/BaseCheck.java

```diff
@@ -364,6 +367,37 @@ protected CheckFlag createFlag(final AtlasObject object, final List<String> inst
                 instructions, points);
     }
 
+    /**
+     * If this object is a way-sectioned edge, return a set of all edges that originated from the
+     * set. Otherwise, return a singleton containing this edge.
+     *
+     * @param object
+     *            Any AtlasObject
+     * @return A set of objects related to the original OSM data.
+     */
+    protected Set<AtlasObject> gatherWaySectionedEdges(final AtlasObject object)
+    {
+        if (object instanceof Edge)
+        {
+            return Collections.unmodifiableSet(new OsmWayWalker((Edge) object).collectEdges());
```

Comment:
Why an unmodifiableSet?

Reply:
Good question! This was something I was struggling against for a bit. What it boils down to is this:
 - `Walker.collectEdges()` returns a `Set<Edge>`.
 - `createFlag` expects a `Set<AtlasObject>`. Note that `Set<Edge>` isn't a valid argument -- consider the following functions:
```
    private void tricky(final Set<AtlasObject> objects) {
        objects.add(new Point(0,0));
    }

    private void whoops(final Edge anEdge) {
        final Set<Edge> set = new HashSet<>();
        set.add(anEdge);
        tricky(set); // won't compile
        set.stream().forEach(edge -> doSomeEdgeSpecificStuff); // this would be bad!
    }
```
Possible solutions to this are:
 - Copy each element from `collectEdges()` into a new set. This isn't particularly efficient but also requires few changes to the existing code.
 - Make `createFlag` accept `Set<? extends AtlasObject>`. This has some cascading effects but I _think_ it's not too bad. Ultimately this may be the best approach, but it definitely has the potential to be a breaking change downstream so we should be careful with it.
 - Call `Collections.unmodifiableSet()`, which a) wraps an existing set instead of copying it to a new set b) guarantees type-safety by throwing an error if anyone tries to add anything. I was inspired by [this stackoverflow post](https://stackoverflow.com/questions/5082044/most-efficient-way-to-cast-listsubclass-to-listbaseclass).

I decided to g

Reply:
You can see all the things that had to change to make the second option compile in 2775dfb if you're curious.

Reply:
Ok, that is quite legit. I agree that ultimately it would be great to implement the second option, but that should probably have its own PR if and when it is decided that should be done.

Reply:
That's how I feel too!

Reply:
Since we ended up going with the second approach, I'm going to mark this as resolved!